ui/src/app/components/popups.tsx

```diff
@@ -114,16 +114,50 @@ export const computePopupPosition = ({side, viewport, target, element, gap}) =>
       ['top', () => ({top: target.top - element.height - gap, left})],
       ['bottom', () => ({top: target.bottom + gap, left})],
       ['left', () => ({left: target.left - element.width - gap, top})],
-      ['right', () => ({left: target.right + gap, top})]
+      ['right', () => ({left: target.right + gap, top})],
+      // Using half a gap here so the popup lines up with the snowman menu.
+      ['bottom-left', () => ({top: target.bottom + gap, left: target.left - element.width + (.5 * gap)})],
+      ['bottom-right', () => ({top: target.bottom + gap, left: target.right - (.5 * gap)})],
+      ['top-left', () => ({top: target.top - element.height - gap, left: target.left - element.width + (.5 * gap)})],
+      ['top-right', () => ({top: target.top - element.height - gap, left: target.right - (.5 * gap)})]
     );
   };
   const position = getPosition(side);
+  const overflowsTop = position.top < 0;
+  const overflowsBottom = position.top + element.height >= viewport.height;
+  const overflowsLeft = position.left < 0;
+  const overflowsRight = position.left + element.width >= viewport.width;
   const maybeFlip = d => {
     return switchCase(d,
-      ['top', () => position.top < 0 ? 'bottom' : 'top'],
-      ['bottom', () => position.top + element.height >= viewport.height ? 'top' : 'bottom'],
-      ['left', () => position.left < 0 ? 'right' : 'left'],
-      ['right', () => position.left + element.width >= viewport.width ? 'left' : 'right']
+      ['top', () => overflowsTop ? 'bottom' : 'top'],
+      ['bottom', () => overflowsBottom ? 'top' : 'bottom'],
+      ['left', () => overflowsLeft ? 'right' : 'left'],
+      ['right', () => overflowsRight ? 'left' : 'right'],
+      // Unfortunately, nested switchCase is probably the clearest way to express this.
+      ['bottom-left', () => switchCase({ob: overflowsBottom, ol: overflowsLeft},
+          [{ob: false, ol: false}, () => 'bottom-left'],
```

Comment:
Thanks, this is definitely better - and nice that the `overflowsX` computations are pulled out and labeled now.

Not blocking, but curious if @petesantos has any other suggestions on this. I don't love the true/false in the object literal here, but I don't have better ideas from our current utilities, or from lodash.

Reply:
Rather than enumerate all of the states in a `switchCase` we could "build" the new positional information:

```javascript
    // This should be exported from utils
    // Will return the result of the first function if not falsey, otherwise returns the result of the second
    const either = fp.curry((fnA, fnB) => (...args) => fnA(...args) || fnB(...args))
    
    // Passes the position information through if it is not the relevant side
    const maybePassPosition = fp.curry((side, position, result) => fp.includes(side, position) ? null : result)

    const maybeOverflowsTop = (result = []) => [...result, position.top < 0 ? 'bottom' : 'top']
    const maybeOverflowsLeft = (result = []) => [...result, position.left < 0 ? 'right': 'left']
    const maybeOverflowsBottom = (result = []) => 
        [...result, position.top + element.height >= viewport.height ? 'top' : 'bottom'];
    const maybeOverflowsRight = (result = []) => 
        [...result,  position.left + element.width >= viewport.width ? 'left' : 'right'];

    const maybeFlip = position => fp.flow(
        either(maybePassPosition('top', position), maybeOverflowsTop),
        either(maybePassPosition('bottom', position), maybeOverflowsRight),
        either(maybePassPosition('left', position), maybeOverflowsLeft),
        either(maybePassPosition('right', position), maybeOverflowsBottom),
        fp.join('-')
    )([])
    console.log('top-left:', maybeFlip('top-left') )
```

Reply:
That looks really nice! It is much clearer and easier to understand. Mapping over the `switchCase` really takes advantage of the fact that the `switchCase` is an expression rather than a standard case statement.